netenv/location.go

```diff
@@ -2,86 +2,265 @@ package netenv
-	// TODO: Create IPv6 version of GetApproximateInternetLocation
+type DeviceLocations struct {
+	Best *DeviceLocation
+	All  map[string]*DeviceLocation
+}
+
+func copyDeviceLocations() *DeviceLocations {
+	locationsLock.Lock()
+	defer locationsLock.Unlock()
+
+	// Create a copy of the locations, but not the entries.
+	cp := *locations
+	cp.All = make(map[string]*DeviceLocation, len(locations.All))
+	for k, v := range locations.All {
+		cp.All[k] = v
+	}
+
+	return &cp
+}
+
+// DeviceLocation represents a single IP and metadata. It must not be changed
+// once created.
+type DeviceLocation struct {
+	IP             net.IP
+	Continent      string
+	Country        string
+	ASN            uint
+	ASOrg          string
+	Source         DeviceLocationSource
+	SourceAccuracy int
+}
+
+type DeviceLocationSource string
+
+const (
+	SourceInterface  DeviceLocationSource = "interface"
+	SourcePeer       DeviceLocationSource = "peer"
+	SourceUPNP       DeviceLocationSource = "upnp"
+	SourceTraceroute DeviceLocationSource = "traceroute"
+	SourceOther      DeviceLocationSource = "other"
+)
+
+func (dls DeviceLocationSource) Accuracy() int {
+	switch dls {
+	case SourceInterface:
+		return 5
+	case SourcePeer:
+		return 4
+	case SourceUPNP:
+		return 3
+	case SourceTraceroute:
+		return 2
+	case SourceOther:
+		return 1
+	default:
+		return 0
+	}
+}
+
+func SetInternetLocation(ip net.IP, source DeviceLocationSource) (ok bool) {
+	// Check if IP is global.
+	if netutils.GetIPScope(ip) != netutils.Global {
+		return false
+	}
+
+	// Create new location.
+	loc := &DeviceLocation{
+		IP:             ip,
+		Source:         source,
+		SourceAccuracy: source.Accuracy(),
+	}
+
+	locationsLock.Lock()
+	defer locationsLock.Unlock()
+
+	// Add to locations, if better.
+	key := loc.IP.String()
+	existing, ok := locations.All[key]
+	if ok && existing.SourceAccuracy > loc.SourceAccuracy {
+		// Existing entry is better.
+		// Return true, because the IP address is part of the locations.
+		return true
+	}
+	locations.All[key] = loc
 
-	// First check if we have an assigned IPv6 address. Return that if available.
-	globalIPv4, _, err := GetAssignedGlobalAddresses()
+	// Find best location.
+	var best *DeviceLocation
+	for _, dl := range locations.All {
+		if best == nil || dl.SourceAccuracy > best.SourceAccuracy {
+			best = dl
+		}
+	}
+	locations.Best = best
+
+	// Get geoip information, but continue if it fails.
+	geoLoc, err := geoip.GetLocation(ip)
 	if err != nil {
-		log.Warningf("netenv: location approximation: failed to get assigned global addresses: %s", err)
-	} else if len(globalIPv4) > 0 {
-		return globalIPv4[0], nil
+		log.Warningf("netenv: failed to get geolocation data of %s (from %s): %s", ip, source, err)
+	} else {
+		loc.Continent = geoLoc.Continent.Code
+		loc.Country = geoLoc.Country.ISOCode
+		loc.ASN = geoLoc.AutonomousSystemNumber
+		loc.ASOrg = geoLoc.AutonomousSystemOrganization
+	}
```

Comment:
This is not related to the `locations` variable and can be fetched outside the `locationsLock`.